src/Sarif/Core/WebMessageUtilities.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    /// <summary>
+    /// This class provides utilities common to the parsing of <see cref="WebRequest"/>
+    /// and <see cref="WebResponse"/> objects from the raw request and response strings.
+    /// </summary>
+    /// <remarks>
+    /// Ideally, this would have been a base class for both WebRequest and WebResponse.
+    /// Unfortunately, every class in the SARIF SDK's object model derives from
+    /// <see cref="PropertyBagHolder"/>, so the .NET class inheritance point is already
+    /// spent.
+    /// </remarks>
+    internal static class WebMessageUtilities
+    {
+        // When including the contents of a web request or response in an exception message, truncate
+        // it to this length:
+        private const int MaxExceptionMessageStringLength = 200;
+
+        internal const string CRLF = "\r\n";
+        internal const string TokenPattern = "[!#$%&'*+._`|~0-9a-zA-Z^-]+";
```

Comment:
>internal [](start = 8, length = 8)

The patterns are based on the grammar in RFC 7230, "Hypertext Transfer Protocol (HTTP/1.1): Message Syntax and Routing".

Reply:
thanks! could we mention so in a comment in code?

---
In reply to: [295491099](https://github.com/microsoft/sarif-sdk/pull/1537#discussion_r295491099) [](ancestors = 295491099)

Reply:
Yes, will do. In fact whenever I add a PR comment I ask myself "Would this be better as a code comment?" -- and I missed this one. Good find.

---
In reply to: [295567359](https://github.com/microsoft/sarif-sdk/pull/1537#discussion_r295567359) [](ancestors = 295567359,295491099)

Reply:
```
        // These patterns are taken from the grammar defined in RFC 7230, "Hypertext Transfer Protocol (HTTP/1.1):
        // Message Syntax and Routing". Yes, tokens (such as message header field names) really can have all those
        // weird characters.
        //
        // We are a little looser than the RFC in a couple of places. In the request, we don't verify that the
        // "target" field on the request line is a valid URI. Nor do we verify that the request body or response
        // body contains only "VCHAR"s (visible characters as opposed to control characters), as required by
        // the RFC.
```

---
In reply to: [295568281](https://github.com/microsoft/sarif-sdk/pull/1537#discussion_r295568281) [](ancestors = 295568281,295567359,295491099)